File: sdk/python/openclaw_managed_agents/resources/sessions.py
"""Session resource methods."""

from __future__ import annotations

import json
from typing import Any, Dict, Iterator, List, Optional

import httpx
from httpx_sse import connect_sse

from ..types import Event, Session


def _parse_session(data: Dict[str, Any]) -> Session:
    return Session(
        session_id=data["session_id"],
        agent_id=data["agent_id"],
        status=data["status"],
        tokens=data.get("tokens", {"input": 0, "output": 0}),
        cost_usd=data.get("cost_usd", 0),
        created_at=data["created_at"],
        output=data.get("output"),
        environment_id=data.get("environment_id"),
        error=data.get("error"),
        last_event_at=data.get("last_event_at"),
    )


def _parse_event(data: Dict[str, Any]) -> Event:
    return Event(
        event_id=data["event_id"],
        session_id=data["session_id"],
        type=data["type"],
        content=data["content"],
        created_at=data["created_at"],
        tokens=data.get("tokens"),
        cost_usd=data.get("cost_usd"),
        model=data.get("model"),
        tool_name=data.get("tool_name"),
        tool_call_id=data.get("tool_call_id"),
        tool_arguments=data.get("tool_arguments"),
        is_error=data.get("is_error"),
        approval_id=data.get("approval_id"),
    )


class Sessions:
    def __init__(self, client: httpx.Client) -> None:
        self._client = client

    def create(
        self,
        *,
        agent_id: str,
        environment_id: Optional[str] = None,
    ) -> Session:
        body: Dict[str, Any] = {"agentId": agent_id}
        if environment_id is not None:
            body["environmentId"] = environment_id
        resp = self._client.post("/v1/sessions", json=body)
        resp.raise_for_status()
        return _parse_session(resp.json())

    def get(self, session_id: str) -> Session:
        resp = self._client.get(f"/v1/sessions/{session_id}")
        resp.raise_for_status()
        return _parse_session(resp.json())

    def list(self) -> List[Session]:
        resp = self._client.get("/v1/sessions")
        resp.raise_for_status()
        return [_parse_session(s) for s in resp.json()["sessions"]]

    def delete(self, session_id: str) -> None:
        resp = self._client.delete(f"/v1/sessions/{session_id}")
        resp.raise_for_status()

    def send(
        self,
        session_id: str,
        *,
        content: str,
        model: Optional[str] = None,
    ) -> Dict[str, Any]:
        """Post a user message to a session. Returns session status and queued flag."""
        body: Dict[str, Any] = {"content": content}
        if model is not None:
            body["model"] = model
        resp = self._client.post(f"/v1/sessions/{session_id}/events", json=body)
        resp.raise_for_status()
        return resp.json()

    def confirm_tool(
        self,
        session_id: str,
        *,
        tool_use_id: str,
        result: str,
        deny_message: Optional[str] = None,
    ) -> Dict[str, Any]:
        """Resolve a pending tool confirmation (always_ask policy)."""
        body: Dict[str, Any] = {
            "type": "user.tool_confirmation",
            "toolUseId": tool_use_id,
            "result": result,
        }
        if deny_message is not None:
            body["denyMessage"] = deny_message
        resp = self._client.post(f"/v1/sessions/{session_id}/events", json=body)
        resp.raise_for_status()
        return resp.json()

    def cancel(self, session_id: str) -> Dict[str, Any]:
        """Cancel the in-flight run on a session."""
        resp = self._client.post(f"/v1/sessions/{session_id}/cancel")
        resp.raise_for_status()
        return resp.json()

    def events(self, session_id: str) -> List[Event]:
        """Get the full event history for a session."""
        resp = self._client.get(f"/v1/sessions/{session_id}/events")
        resp.raise_for_status()
        return [_parse_event(e) for e in resp.json()["events"]]

    def stream(self, session_id: str) -> Iterator[Event]:
        """SSE stream of events. Catches up on existing events then tail-follows.

        Yields Event objects. Automatically skips heartbeat events.
        The iterator ends when the server closes the connection (session
        idle for ~30s after the last event).

        Usage::

            for event in client.sessions.stream(session_id):
                if event.type == "agent.message":
                    print(event.content)
        """
        with connect_sse(
            self._client,
            "GET",
            f"/v1/sessions/{session_id}/events",
            params={"stream": "true"},
        ) as sse:
            for server_event in sse.iter_sse():
                if server_event.event == "heartbeat":
                    continue
                try:
                    data = json.loads(server_event.data)
                except (json.JSONDecodeError, TypeError):
                    continue
                yield _parse_event(data)

File: sdk/typescript/src/resources/sessions.ts
import type { HttpClient } from "../http.js";
import { parseSse } from "../sse.js";
import type {
  CancelResult,
  CompactResult,
  Event,
  SendEventResult,
  Session,
  ThinkingLevel,
} from "../types.js";

export interface CreateSessionParams {
  agentId: string;
  environmentId?: string;
  vaultId?: string;
}

export interface SendParams {
  content: string;
  model?: string;
  thinkingLevel?: ThinkingLevel;
}

export interface ConfirmToolParams {
  toolUseId: string;
  result: "allow" | "deny";
  denyMessage?: string;
}

export class Sessions {
  constructor(private readonly http: HttpClient) {}

  create(params: CreateSessionParams): Promise<Session> {
    const body: Record<string, unknown> = { agentId: params.agentId };
    if (params.environmentId !== undefined) body["environmentId"] = params.environmentId;
    if (params.vaultId !== undefined) body["vaultId"] = params.vaultId;
    return this.http.request<Session>("POST", "/v1/sessions", body);
  }

  get(sessionId: string): Promise<Session> {
    return this.http.request<Session>("GET", `/v1/sessions/${encodeURIComponent(sessionId)}`);
  }

  async list(): Promise<Session[]> {
    const resp = await this.http.request<{ sessions: Session[] }>("GET", "/v1/sessions");
    return resp.sessions;
  }

  async delete(sessionId: string): Promise<void> {
    await this.http.request<void>(
      "DELETE",
      `/v1/sessions/${encodeURIComponent(sessionId)}`,
    );
  }

  send(sessionId: string, params: SendParams): Promise<SendEventResult> {
    const body: Record<string, unknown> = { content: params.content };
    if (params.model !== undefined) body["model"] = params.model;
    if (params.thinkingLevel !== undefined) body["thinkingLevel"] = params.thinkingLevel;
    return this.http.request<SendEventResult>(
      "POST",
      `/v1/sessions/${encodeURIComponent(sessionId)}/events`,
      body,
    );
  }

  confirmTool(sessionId: string, params: ConfirmToolParams): Promise<SendEventResult> {
    const body: Record<string, unknown> = {
      type: "user.tool_confirmation",
      toolUseId: params.toolUseId,
      result: params.result,
    };
    if (params.denyMessage !== undefined) body["denyMessage"] = params.denyMessage;
    return this.http.request<SendEventResult>(
      "POST",
      `/v1/sessions/${encodeURIComponent(sessionId)}/events`,
      body,
    );
  }

  cancel(sessionId: string): Promise<CancelResult> {
    return this.http.request<CancelResult>(
      "POST",
      `/v1/sessions/${encodeURIComponent(sessionId)}/cancel`,
    );
  }

  compact(sessionId: string): Promise<CompactResult> {
    return this.http.request<CompactResult>(
      "POST",
      `/v1/sessions/${encodeURIComponent(sessionId)}/compact`,
    );
  }

  logs(sessionId: string, params: { tail?: number } = {}): Promise<string> {
    const qs = params.tail === undefined ? "" : `?tail=${encodeURIComponent(String(params.tail))}`;
    return this.http.textRequest(
      "GET",
      `/v1/sessions/${encodeURIComponent(sessionId)}/logs${qs}`,
    );
  }

  async events(sessionId: string): Promise<Event[]> {
    const resp = await this.http.request<{ events: Event[] }>(
      "GET",
      `/v1/sessions/${encodeURIComponent(sessionId)}/events`,
    );
    return resp.events;
  }

  /**
   * SSE stream of events. Catches up on existing events, then tail-follows.
   * Skips heartbeats. The iterator ends when the server closes the connection
   * (session has been idle for ~30s with no new events).
   *
   * Example:
   * ```ts
   * for await (const event of client.sessions.stream(sessionId)) {
   *   if (event.type === "agent.message") console.log(event.content);
   * }
   * ```
   */
  async *stream(sessionId: string): AsyncGenerator<Event> {
    const resp = await this.http.streamRequest(
      `/v1/sessions/${encodeURIComponent(sessionId)}/events?stream=true`,
    );
    for await (const sse of parseSse(resp)) {
      if (sse.event === "heartbeat") continue;
      if (sse.data.length === 0) continue;
      let parsed: unknown;
      try {
        parsed = JSON.parse(sse.data);
      } catch {
        continue;
      }
      yield parsed as Event;
    }
  }
}
